File: 733-flood-fill/flood-fill.cpp
class Solution {
public:
    int m, n;
    vector<vector<int>> direction{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

    void dfs(vector<vector<int>>& ans, int r, int c, int intialcolor,
             int color) {
        if (r < 0 || r >= m || c < 0 || c >= n || ans[r][c] != intialcolor)
            return;

        ans[r][c] = color;

        for (auto dir : direction) {
            int x = r + dir[0];
            int y = c + dir[1];
            dfs(ans, x, y, intialcolor, color);
        }
    }
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc,
                                  int color) {
        vector<vector<int>> ans = image;

        m = ans.size();
        n = ans[0].size();

        int intialcolor = ans[sr][sc];

        if (intialcolor != color) {
            dfs(ans, sr, sc, intialcolor, color);
        }
        return ans;
    }
};

File: 733-flood-fill/flood-fill.java
public class Solution {

    int m, n;

    int[][] dirs = {
        {1, 0}, {-1, 0}, {0, 1}, {0, -1}
    };


    private void dfs(int[][] image, int r, int c, int origColor, int newColor) {
      
        if (r < 0 || r >= m || c < 0 || c >= n || image[r][c] != origColor) {
            return;
        }

        
        image[r][c] = newColor;

        
        for (int[] d : dirs) {
            dfs(image, r + d[0], c + d[1], origColor, newColor);
        }
    }



    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        m = image.length;
        n = image[0].length;
        int origColor = image[sr][sc];
        

        if (origColor != newColor) {
            dfs(image, sr, sc, origColor, newColor);
        }
        return image;
    }
}